qadence/ml_tools/train_grad.py

```diff
@@ -160,115 +160,167 @@ def loss_fn(model: torch.nn.Module, data: torch.Tensor) -> tuple[torch.Tensor, d
+        xs = next(dl_iter) if dl_iter is not None else None
+        xs_to_device = data_to_device(xs, device=device, dtype=data_dtype)
+        return loss_fn(model, xs_to_device)
+
+    # populate callbacks with already available internal functions
+    # printing, writing and plotting
+    callbacks = config.callbacks
+
+    # printing
+    if config.verbose and config.print_every > 0:
+        # Note that the loss returned by optimize_step
+        # is the value before doing the training step
+        # which is printed accordingly by the previous iteration number
+        callbacks += [
+            Callback(
+                lambda opt_res: print_metrics(opt_res.loss, opt_res.metrics, opt_res.iteration - 1),
+                every=config.print_every,
+                call_after_opt=True,
+            )
+        ]
+
+    # plotting
+    callbacks += [
+        Callback(
+            lambda opt_res: plot_tracker(
+                writer,
+                opt_res.model,
+                opt_res.iteration,
+                config.plotting_functions,
+                tracking_tool=config.tracking_tool,
+            ),
+            every=config.plot_every,
+            call_before_opt=True,
+        )
+    ]
+
+    # writing metrics
+    callbacks += [
+        Callback(
+            lambda opt_res: write_tracker(
+                writer,
+                opt_res.loss,
+                opt_res.metrics,
+                opt_res.iteration,
+                tracking_tool=config.tracking_tool,
+            ),
+            every=config.write_every,
+            call_before_opt=False,
+            call_after_opt=True,
+            call_during_eval=True,
+        )
+    ]
+
+    # checkpointing
+    if config.folder and config.checkpoint_every > 0 and not config.checkpoint_best_only:
+        callbacks += [
+            Callback(
+                lambda opt_res: write_checkpoint(
+                    config.folder,  # type: ignore[arg-type]
+                    opt_res.model,
+                    opt_res.optimizer,
+                    opt_res.iteration,
+                ),
+                every=config.checkpoint_every,
+                call_before_opt=False,
+                call_after_opt=True,
+            )
+        ]
+
+    if config.folder and config.checkpoint_best_only:
+        callbacks += [
+            Callback(
+                lambda opt_res: write_checkpoint(
+                    config.folder,  # type: ignore[arg-type]
+                    opt_res.model,
+                    opt_res.optimizer,
+                    "best",
+                ),
+                every=config.checkpoint_every,
+                call_before_opt=True,
+                call_after_opt=True,
+                call_during_eval=True,
+            )
+        ]
+
+    def run_callbacks(callback_iterable: list[Callback], opt_res: OptimizeResult) -> None:
+        [callback(opt_res) for callback in callback_iterable]
```

Comment:
why the list comprehension?

Reply:
Just liked one-liners but putting it back to a simple loop syntax